Rakefile

```diff
@@ -48,16 +48,60 @@ task :check_for_node do
 end
 
 desc "Update JavaScript components"
-task :update_js => :check_for_node do
+task :update_js_components => :check_for_node do
   abort unless system('bower', 'update')
   abort unless system('bower-installer')
 end
 
-desc "Compress JavaScript components"
-task :compress_js => :check_for_node do
-  Dir[File.join(File.dirname(__FILE__), 'assets', 'js', '**', '*')].each do |f|
-    next unless File.file? f and f.end_with? '.js' and not f.end_with? '.min.js'
-    outfile = "#{f[0..-('.js'.size + 1)]}.min.js"
-    abort unless system('uglifyjs', '-c', '-m', '--', f, :out=>outfile)
+LIB_LUNR_TARGET = File.join(%w(lib jekyll_pages_api_search lunr.min.js))
+LIB_LUNR_SOURCE = File.join(%w(assets js vendor lunr.js lunr.js))
+
+file LIB_LUNR_TARGET => LIB_LUNR_SOURCE do
+  abort unless system('uglifyjs', '-c', '-m', '-o', LIB_LUNR_TARGET,
+    '--', LIB_LUNR_SOURCE)
+end
+
+# The following parses the build.js used by the RequireJS package's r.js
+# optimization tool. The `mainConfigFile` member from that file is parsed to
+# discover all of the JavaScript files that the `out` member depends on.
+# TODO(mbland): Extract this and other bits from this Rakefile into a gem.
+cxt = V8::Context.new
+build_js = nil
+basedir = File.dirname(__FILE__)
+File.open(File.join(basedir, 'build.js')) {|f| build_js = f.read}
+cxt.eval "var build_js = #{build_js};"
+config = cxt[:build_js].mainConfigFile
+config_dir = File.dirname(config)
+requirejs_outfile = cxt[:build_js].out
+cxt.load File.join(basedir, 'require-shim.js')
+cxt.load config
+requirejs_paths = []
+requirejs_config = cxt[:requirejs_config]
+requirejs_config.paths.values.each {|i| requirejs_paths << i}
+requirejs_config.deps.each {|i| requirejs_paths << i}
+requirejs_paths = requirejs_paths.map {|i| "#{File.join config_dir, i}.js"}
+
+file requirejs_outfile => requirejs_paths do
+  abort unless system('r.js', '-o', 'build.js')
+end
+
+requirejs_outfile_gz = "#{requirejs_outfile}.gz"
+
+file requirejs_outfile_gz => requirejs_outfile do
+  unless program_exists? 'gzip'
+    puts "Cannot determine if the gzip program exists on the system; " +
+      "skipping compression."
+    return
+  end
+  unless system('gzip', '--best', '-c', requirejs_outfile,
+    :out=>requirejs_outfile_gz)
+    abort "compression failed for: #{requrejs_outfile}"
   end
 end
```

Comment:
Doesn't https://github.com/jekyll-assets/jekyll-assets do a lot of this? Also, if we're requiring Node to do the build anyway, wonder if we'd be better off managing all of this in one of their build tools. We'd get a lot of this for free.

http://blog.cozycloud.cc/technic/2014/06/18/task-runners-comparison/